### PR TITLE
[Editor] Guess font size and color from the AS of FreeText annotations

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: src/core/annotation.js [hunk 1]
```diff
@@ -3540,28 +3541,34 @@ class FreeTextAnnotation extends MarkupAnnotation {
   constructor(params) {
     super(params);
 
-    this.data.hasOwnCanvas = this.data.noRotate;
+    this.data.hasOwnCanvas = true;
```
Comment (Snuffleupagus) [inline src/core/annotation.js:3544]: How is this related to the other changes?

File: src/core/default_appearance.js [hunk 2]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
   return new DefaultAppearanceEvaluator(str).parse();
 }
 
+class AppearanceEvaluator extends EvaluatorPreprocessor {
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:90]: To reduce confusion with the pre-existing class, please use a slightly more specific name:
```suggestion
class AppearanceStreamEvaluator extends EvaluatorPreprocessor {
```

File: src/core/default_appearance.js [hunk 3]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
...
+      fn: 0,
+      args: [],
+    };
+    const result = {
+      fontSize: 0,
+      fontName: "",
+      fontColor: /* black = */ new Uint8ClampedArray(3),
+    };
+
+    let isInTextSection = false;
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:107]: Given that consistency is nice, please let's use the same name as we do in the regular evaluator:
```suggestion
    let parsingText = false;
```

File: src/core/default_appearance.js [hunk 4]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
   return new DefaultAppearanceEvaluator(str).parse();
 }
 
+class AppearanceEvaluator extends EvaluatorPreprocessor {
+  constructor(str) {
+    super(str);
+    this.str = str;
+  }
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:94]: Let's be explicit about this method expecting a *stream*, since we most of the time use "str" as a shorthand for "string" in the code-base:
```suggestion
  constructor(stream) {
    super(stream);
    this.stream = stream;
  }
```

File: src/core/default_appearance.js [hunk 5]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
...
+          case OPS.setFillColorSpace:
+            if (!isInTextSection) {
+              break;
+            }
+            ColorSpace.singletons.cmyk.getRgbItem(args, 0, result.fontColor, 0);
+            break;
+        }
+      }
+    } catch (reason) {
+      warn(`parseDefaultAppearance - ignoring errors: "${reason}".`);
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:163]: Too much copy-paste here:
```suggestion
      warn(`parseAppearanceStream - ignoring errors: "${reason}".`);
```

File: src/core/default_appearance.js [hunk 6]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
...
+    this.str.reset();
+
+    return result;
+  }
+}
+
+// Parse appearance stream to extract font and color information.
+function parseAppearanceStream(str) {
+  return new AppearanceEvaluator(
+    typeof str === "string" ? new StringStream(str) : str
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:174]: Shouldn't the appearance *always* be a Stream, since I believe that the code in `src/core/annotation.js` ensures that?

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: src/core/default_appearance.js [hunk 7]
```diff
@@ -87,6 +87,94 @@ function parseDefaultAppearance(str) {
...
+        switch (fn | 0) {
+          case OPS.beginText:
+            isInTextSection = true;
+            break;
+          case OPS.endText:
+            isInTextSection = false;
+            scaleFactor = 1;
+            break;
+          case OPS.setTextMatrix:
+            scaleFactor *= Math.abs(args[0]);
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:128]: Also, isn't it possible that other components of the textMatrix also affects the size?

File: src/core/annotation.js [hunk 8]
```diff
@@ -3540,28 +3541,34 @@ class FreeTextAnnotation extends MarkupAnnotation {
...
         this.rectangle,
         this.rotation,
-        fontData.fontSize || 10,
-        fontData.fontColor,
+        fontSize || 10,
+        fontColor,
         strokeAlpha
       );
+      this.data.fontColor = Array.from(fontColor);
+      this.data.fontSize = fontSize || 10;
```
Comment (Snuffleupagus) [inline src/core/annotation.js:3562]: Also, here and below, would it not be much better (for API-consistency) to override the existing `this.data.defaultAppearanceData` data with these new values rather than adding a bunch of new properties?

File: src/core/annotation.js [hunk 9]
```diff
@@ -3540,28 +3541,34 @@ class FreeTextAnnotation extends MarkupAnnotation {
...
       this._streams.push(this.appearance, FakeUnicodeFont.toUnicodeStream);
     } else if (!this._isOffscreenCanvasSupported) {
       warn(
         "FreeTextAnnotation: OffscreenCanvas is not supported, annotation may not render correctly."
       );
+    } else if (this.appearance) {
+      const { fontColor, fontSize } = parseAppearanceStream(this.appearance);
+      this.data.fontColor = Array.from(fontColor);
+      this.data.fontSize = fontSize;
     }
```
Comment (Snuffleupagus) [inline src/core/annotation.js:3572]: Are all of these `else if` branches actually working as intended now?
Would it not be more correct to reformat [all this code](https://github.com/mozilla/pdf.js/pull/16522/files#diff-a7e409abd3d3209d32d3adc8457c5e23feb4c06113a2e065b0c6a9322a3a3369R3549-R3572) as follows:
```javascript
if (this.appearance) {
  const { fontColor, fontSize } = parseAppearanceStream(this.appearance);
  ...
} else if (this._isOffscreenCanvasSupported) {
  const strokeAlpha = params.dict.get("CA");
  ...
} else {
  warn(
    "FreeTextAnnotation: OffscreenCanvas is not supported, annotation may not render correctly."
  );
```

File: src/core/default_appearance.js [hunk 10]
```diff
@@ -87,6 +87,83 @@ function parseDefaultAppearance(str) {
...
+            parsingText = false;
+            scaleFactor = 1;
+            break;
+          case OPS.setTextMatrix:
+            scaleFactor *= Math.hypot(args[0], args[1]);
+            break;
+          case OPS.setFont:
+            if (!parsingText) {
+              break;
+            }
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:133]: Do we actually want to ignore setFont *outside* of the beginText/endText operators, since we don't do that in `DefaultAppearanceEvaluator` above?

Comment (calixteman) [reply src/core/default_appearance.js:133]: The DA isn't really supposed to contain BT/ET but just the minimal commands to set the style.
It probably doesn't hurt to get the font info wherever they're.

File: test/unit/default_appearance_spec.js [hunk 11]
```diff
@@ -50,4 +52,148 @@ describe("Default appearance", function () {
       });
     });
   });
+
+  describe("parseAppearanceStream", () => {
+    it("should parse a FreeText (from Acrobat) appearance", () => {
+      const appearance = `
```
Comment (Snuffleupagus) [inline test/unit/default_appearance_spec.js:58]: Here, and in the tests below, I'd suggest defining this using the "correct" format since that makes it clearer what the correct data is.
```suggestion
      const appearance = new StringStream(`
```

File: test/unit/default_appearance_spec.js [hunk 12]
```diff
@@ -50,4 +52,148 @@ describe("Default appearance", function () {
...
+      ET
+      Q`;
+      const result = {
+        fontSize: 14,
+        fontName: "Helv",
+        fontColor: new Uint8ClampedArray([107, 217, 41]),
+      };
+      expect(parseAppearanceStream(new StringStream(appearance))).toEqual(
+        result
+      );
```
Comment (Snuffleupagus) [inline test/unit/default_appearance_spec.js:89]: With the comment above implemented, we can also ensure that the stream is being reset as intended:
```suggestion
      );
      expect(appearance.pos).toEqual(0);
```

File: src/core/default_appearance.js [hunk 13]
```diff
@@ -87,6 +87,75 @@ function parseDefaultAppearance(str) {
...
+      fn: 0,
+      args: [],
+    };
+    const result = {
+      fontSize: 0,
+      fontName: "",
+      fontColor: /* black = */ new Uint8ClampedArray(3),
+    };
+
+    let scaleFactor = 1;
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:107]: Sorry to complicate things further here, but shouldn't this be reset to a *previous* value on `OPS.restore`?

Comment (calixteman) [reply src/core/default_appearance.js:108]: That said it's strange to ship something which is know to be buggy... Let's fix the save/restore case.

File: src/core/default_appearance.js [hunk 14]
```diff
@@ -87,6 +87,91 @@ function parseDefaultAppearance(str) {
...
+    const stack = [];
+
+    let scaleFactor = 1;
+    let mustLoop = true;
+
+    try {
+      while (mustLoop) {
+        operation.args.length = 0; // Ensure that `args` it's always reset.
+
+        if (!this.read(operation)) {
```
Comment (Snuffleupagus) [inline src/core/default_appearance.js:115]: To improve overall consistency, I'd probably do the following instead:
```suggestion
    let breakLoop = false;

    try {
      while (true) {
        operation.args.length = 0; // Ensure that `args` it's always reset.

        if (breakLoop || !this.read(operation)) {
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with a couple more comments and passing tests; thank you!
